sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceExceptWithFilter.scala

```diff
@@ -0,0 +1,114 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+
+/**
+ * If one or both of the datasets in the logical [[Except]] operator are purely transformed using
+ * [[Filter]], this rule will replace logical [[Except]] operator with a [[Filter]] operator by
+ * flipping the filter condition of the right child.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 WHERE a2 = 12 EXCEPT SELECT a1, a2 FROM Tab1 WHERE a1 = 5
+ *   ==>  SELECT DISTINCT a1, a2 FROM Tab1 WHERE a2 = 12 AND (a1 is null OR a1 <> 5)
+ * }}}
+ *
+ * Note:
+ * Before flipping the filter condition of the right node, we should:
+ * 1. Combine all it's [[Filter]].
+ * 2. Apply InferFiltersFromConstraints rule (to support NULL values of the condition).
+ */
+object ReplaceExceptWithFilter extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Except(left: Project, right) if isEligible(left, right) =>
+      Project(left.projectList,
+        Distinct(Filter(Not(transformCondition(left.child, skipProject(right))), left.child)))
+
+    case Except(left, right) if isEligible(left, right) =>
+      Distinct(Filter(Not(transformCondition(left, skipProject(right))), left))
+  }
+
+  private def transformCondition(left: LogicalPlan, right: LogicalPlan) = {
+    val filterCondition = InferFiltersFromConstraints(combineFilters(right)
+    ).asInstanceOf[Filter].condition
+
+    val attributeNameMap: Map[String, Attribute] = left.output.map(x => (x.name, x)).toMap
+    val transformedCondition = filterCondition transform { case a : AttributeReference =>
+      attributeNameMap(a.name)
+    }
+
+    transformedCondition
+  }
+
+  private def isEligible(left: LogicalPlan, right: LogicalPlan) = (left, right) match {
+    case (_, right @ (Project(_, _: Filter) | Filter(_, _))) => verifyConditions(left, right)
+    case _ => false
+  }
+
+  private def verifyConditions(left: LogicalPlan, right: LogicalPlan) = {
+    val leftProjectList = projectList(left)
+    val rightProjectList = projectList(right)
+
+    verifyFilterCondition(skipProject(left)) && verifyFilterCondition(skipProject(right)) &&
+      Project(leftProjectList, nonFilterChild(skipProject(left))).sameResult(
+        Project(rightProjectList, nonFilterChild(skipProject(right))))
+  }
+
+  private def verifyFilterCondition(plan: LogicalPlan) = {
```

Comment:
To filter out the cases of `SubqueryExpression `,  you can simply check 
```Scala
SubqueryExpression.hasCorrelatedSubquery(e)
```

@dilipbiswal Could you take a look at this rule?

Reply:
That would be perfect, i'll do that, thanks!

Reply:
Also, is there a way to simplify the function something like `val filterConditions = plan.takeWhile(_.isInstanceOf[Filter])` ?

Reply:
@gatorsmile I have a question. Is the intention to skip Filter nodes containing all kinds of subquery expression (i.e correlated or otherwise) ? If so , then we can't call the hasCorrelatedSubquery method as it only detects the correlated subquery expressions. Can we add a utility method  for this in SubqueryExpression ? We may have future usage for it.
```
def hasSubquery(e: Expression): Boolean = {
    e.find {
      case _: SubqueryExpression => true
      case _ => false
    }.isDefined
  }```

Reply:
I am fine if we do not support any subquery expression for such a rule. 

Maybe you can review the whole rule?

Reply:
@gatorsmile yeah.. going through it..

Reply:
@sathiyapk I suspect there may be a regression in the following query. I am using the same schema as the one used in except.sql.
```
select t1.* from t1, t2 where t1.k = t2.k except select t1.* from t1, t2 where t1.k = t2.k and t1.k != 'one'
```
We are producing more rows using this new rule as we are not applying the distinct on the correct set of columns.
Also for this test case, in transformCondition function, we are encountering collision in attribute names as both side of join producing (k#,v#, k#, v#).  For this case, can we please check if logic to replace the attribute references is correct ?
@gatorsmile fyi.. please help double check.

Reply:
@gatorsmile @sathiyapk Do you think we  should explicitly specify the return types for all the private defs in the rule ? I just find it more readable.

Reply:
@dilipbiswal well found! the reason for the more rows is in the case 1 of the rule, we put the Filter below the project. As @gatorsmile suggested, we should remove the first case then the Filter will be added above the project, then it will be resolved..

Reply:
Regarding the collision in the attribute names, the transformCondition function found the right attributes in this specific example, but i agree we should test it with different corner cases. One more conservative solution to this problem could be to filter out the cases where a projectList contains SubqueryExpression like we do for the Filter condition. In any case, if the function didn't find the right attributes, spark will throw an error..

Reply:
And I don't have any problem in explicitly specifying the return types for all the private defs if @gatorsmile is ok with that.

Reply:
Yeah, please follow what @dilipbiswal suggested if the return type can help code readability.

Reply:
@dilipbiswal One quick update regarding the attribute collision case, actually we didn't encountered any collision in the previous example, because we projected only t1.*. Also it looks like we will never encounter this problem..

Reply:
@sathiyapk Ok.. glad that its not an issue. Can you please check the content of left.output ? I remember seeing collision.  
```
left.output.map(x => (x.name, x)).toMap
```
the toMap operation was removing the collision if i remember it correctly. I just wanted to make sure we are replacing the attributes correctly.

Reply:
@dilipbiswal You are right! I first removed the first case of the rule and then tested for collision, that's why i didn't find any collision. Without removing the first case of the rule, yes there is a collision. Although it looks sufficient now, in order to be careful about any unanticipated cases, i will also add an additional check `left.output.size == left.output.distinct.size` in the verifyCondition function.